orchestra/todos/import_export.py

```diff
@@ -0,0 +1,122 @@
+        _write_template_rows(writer, child, depth + 1)
+
+
+def _upload_csv_to_google(spreadsheet_name, file):
+    service = Service(settings.GOOGLE_P12_PATH,
+                      settings.GOOGLE_SERVICE_EMAIL)
+    sheet = service.insert_file(
+        spreadsheet_name,
+        '',
+        settings.ORCHESTRA_TODO_LIST_TEMPLATE_EXPORT_GDRIVE_FOLDER,
+        'text/csv',
+        'application/vnd.google-apps.spreadsheet',
+        file.name
+    )
+    service.add_permission(sheet['id'], write_with_link_permission)
+    return 'https://docs.google.com/spreadsheets/d/{}'.format(sheet['id'])
+
+
+def export_to_spreadsheet(todo_list_template):
+    """
+    Recursively descend down the to-do tree of a `todo_list_template`
+    and add a row per to-do to a CSV, uploading it to a Google Sheet
+    when done. The first two columns of the sheet capture json-encoded
+    remove_if and skip_if logic. After that, the text in the `i`'th
+    column represents a to-do's title that is `i` levels nested in the
+    to-do tree.
+    """
+    with NamedTemporaryFile(mode='w+', delete=False) as file:
+        writer = csv.writer(file)
+        writer.writerow([REMOVE_IF_HEADER, SKIP_IF_HEADER])
+        _write_template_rows(writer, todo_list_template.todos, 0)
+        file.flush()
+        return _upload_csv_to_google(
+            '{} - {}'.format(todo_list_template.name, timezone.now()),
+            file)
+
+
+def import_from_spreadsheet(todo_list_template, spreadsheet_url, request):
+    """Imports a Google Sheet at `spreadsheet_url` to
+    `todo_list_template.todos`.
+
+    For debugging/provenance purposes, creates a
+    TodoListTemplateImportRecord with the user, todo_list_template,
+    and spreadsheet URL behind the import.
+    """
+    reader = get_google_spreadsheet_as_csv(spreadsheet_url, reader=csv.reader)
+    header = next(reader)
+    if header[:2] != [REMOVE_IF_HEADER, SKIP_IF_HEADER]:
+        raise SpreadsheetImportError('Unexpected header: {}'.format(header))
+    # The `i`'th entry in parent_items is current list of child to-dos
+    # of the `i`-depth parent. We use this to determine which parent
+    # to add a child to when we read to-do in a row with lower
+    # indentation than its parent.
+    parent_items = []
+    todos = None
+    for rowindex, row in enumerate(reader):
+        item = {
+            'id': rowindex,
+            'remove_if': json.loads(row[0] or '[]'),
+            'skip_if': json.loads(row[1] or '[]'),
+            'items': []
+        }
+        nonempty_columns = [(columnindex, text)
+                            for columnindex, text in enumerate(row[2:])
+                            if text]
+        if len(nonempty_columns) == 0:
+            continue
+        elif len(nonempty_columns) > 1:
+            raise SpreadsheetImportError(
+                'More than one text entry in row {}: {}'.format(
+                    rowindex, row))
+
+        nonempty_index = nonempty_columns[0][0]
```

Comment:
is `nonempty_index` the same as `depth` in your export logic?

Reply:
Yup! Added a comment confirming this.